packages/mavis-id-sdk/src/provider.ts

```diff
@@ -10,72 +11,78 @@ import {
   UnauthorizedProviderError,
 } from "viem"
 
+import { AuthorizeOpts } from "./auth"
 import { VIEM_CHAIN_MAPPING } from "./common/chain"
 import { Eip1193EventName, Eip1193Provider, MavisIdRequestSchema } from "./common/eip1193"
 import { ID_ORIGIN_PROD } from "./common/gate"
 import { IdResponse } from "./common/id-response"
+import { getScopesParams, Scope } from "./common/scope"
 import { CommunicateHelper } from "./core/communicate"
 import { personalSign } from "./core/personal-sign"
 import { sendTransaction } from "./core/send-tx"
 import { signTypedDataV4 } from "./core/sign-data"
 import { openPopup } from "./utils/popup"
 import { getStorage, removeStorage, setStorage, STORAGE_ADDRESS_KEY } from "./utils/storage"
-import type { Requires } from "./utils/types"
 import { validateIdAddress } from "./utils/validate-address"
 
-export type MavisIdProviderOpts = {
-  clientId: string
+export type MavisIdWalletOpts = AuthorizeOpts & {
   chainId: number
-  redirectUri?: string
-  gateOrigin?: string
-  rpcUrl?: string
 }
 
-export class MavisIdProvider extends EventEmitter implements Eip1193Provider {
+export class MavisIdWallet extends EventEmitter implements Eip1193Provider {
   private readonly clientId: string
-  private readonly gateOrigin: string
-  private readonly redirectUri?: string
+  private readonly idOrigin: string
+  private readonly redirectUrl: string
+  private readonly scopes: Scope[]
 
   readonly chainId: number
   private address?: Address
 
   private readonly viemClient: Client
   private readonly communicateHelper: CommunicateHelper
 
-  protected constructor(
-    options: Requires<MavisIdProviderOpts, "chainId" | "gateOrigin" | "rpcUrl">,
-  ) {
+  protected constructor(options: MavisIdWalletOpts) {
     super()
 
-    const { clientId, chainId, gateOrigin, rpcUrl, redirectUri } = options
+    const {
+      clientId,
+      idOrigin = ID_ORIGIN_PROD,
+      redirectUrl = window.location.origin,
+      scopes = [],
 
-    this.clientId = clientId
-    this.chainId = chainId
-    this.gateOrigin = gateOrigin
-    this.redirectUri = redirectUri
+      chainId,
+    } = options
+    const chain = VIEM_CHAIN_MAPPING[chainId]
 
+    if (!chain) {
+      const err = new Error(`Chain ${chainId} is not supported.`)
+      throw new ChainDisconnectedError(err)
+    }
     this.viemClient = createClient({
       chain: VIEM_CHAIN_MAPPING[chainId],
-      transport: http(rpcUrl),
+      transport: http(),
     })
 
-    this.communicateHelper = new CommunicateHelper(gateOrigin)
-  }
+    // * add default scopes
+    const newScopes = [...scopes]
+    if (scopes.findIndex(s => s === "openid") === -1) {
+      newScopes.push("openid")
+    }
+    if (scopes.findIndex(s => s === "wallet") === -1) {
+      newScopes.push("wallet")
+    }
```

Comment:
this conditional is a bit hard to read, and since this index is not used later, I suggest changing it to use `includes` instead
additionally, `includes` offers a better performance, btw this difference is not significant with such small data like this

```suggestion
    if (!scopes.includes("openid")) {
      newScopes.push("openid")
    }
    if (!scopes.includes("wallet")) {
      newScopes.push("wallet")
    }
```